lib/galaxy/webapps/galaxy/api/tools.py

```diff
@@ -511,9 +512,14 @@ def _create(self, trans, payload, **kwd):
             target_history = None
 
         # Set up inputs.
-        inputs = payload.get('inputs', {})
-        if not isinstance(inputs, dict):
-            raise exceptions.RequestParameterInvalidException("inputs invalid %s" % inputs)
+        submitted_inputs = payload.get('inputs', {})
+        if not isinstance(submitted_inputs, dict):
+            raise exceptions.RequestParameterInvalidException("inputs invalid %s" % submitted_inputs)
+        try:
+            inputs = {}
+            params_to_incoming(inputs, tool.inputs, submitted_inputs, tool.app)
```

Comment:
Can you add an API test for this ? You could modify https://github.com/mvdbeek/galaxy/blob/a2c83ee7224b181c39ed394be7a866aaf88a46c1/lib/galaxy_test/api/test_tools.py#L1550 for instance. I am also thinking that we might want to hide this behind a parameter or a new endpoint. The parameter could be called `legacy_config_form_payload` and default to `True`. Or we add a `api/tools/v2` endpoint that just adds this parameter.

Reply:
Cannot we safely guess the format of `submitted_inputs` by checking if any key contains `|` ?

Reply:
I'm not sure we can, if there is no `|` in an incoming payload it could be the old format, but there's no conditional, or it could be the new format with nested dictionaries. And I prefer to keep this separate, since we probably want to take a different route through the code in the future.